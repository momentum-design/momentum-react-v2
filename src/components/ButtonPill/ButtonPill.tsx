import React, { forwardRef, RefObject } from 'react';
import classnames from 'classnames';
import ButtonSimple from '../ButtonSimple';

import { DEFAULTS, STYLE } from './ButtonPill.constants';
import { Props } from './ButtonPill.types';
import './ButtonPill.style.scss';

const ButtonPill = forwardRef((props: Props, providedRef: RefObject<HTMLButtonElement>) => {
  const {
    className,
    color,
    disabled,
    shallowDisabled,
    ghost,
    grown,
    outline,
    size,
    inverted,
    onlyTriggersTooltip,
    ...otherProps
  } = props;

  if (ghost && inverted) {
    console.warn('MRV2: Momentum does not support a ghost inverted ButtonPill.');
  }

  if (onlyTriggersTooltip && (outline || inverted || color || disabled || shallowDisabled)) {
    console.warn(
      'MRV2: Momentum does not support a button that onlyTriggersTooltip and is outline, inverted, has color, or is disabled'
    );
  }

  return (
    <ButtonSimple
      className={classnames(STYLE.wrapper, className)}
      ref={providedRef}
      data-color={color || DEFAULTS.COLOR}
      data-disabled={disabled || DEFAULTS.DISABLED}
      data-shallow-disabled={shallowDisabled || DEFAULTS.SHALLOW_DISABLED}
      aria-disabled={shallowDisabled || DEFAULTS.SHALLOW_DISABLED}
      data-ghost={ghost || onlyTriggersTooltip || DEFAULTS.GHOST}
      data-grown={grown || DEFAULTS.GROWN}
      data-outline={outline || DEFAULTS.OUTLINE}
      data-size={size || DEFAULTS.SIZE}
      data-inverted={inverted || DEFAULTS.INVERTED}
      data-only-triggers-tooltip={onlyTriggersTooltip || DEFAULTS.ONLY_TRIGGERS_TOOLTIP}
      isDisabled={disabled}
      {...otherProps}
    />
  );
});

ButtonPill.displayName = 'ButtonPill';

export default ButtonPill;
