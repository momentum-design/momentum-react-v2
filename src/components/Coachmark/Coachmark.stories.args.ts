import { commonStyles } from '../../storybook/helper.stories.argtypes';

const coachmarkArgTypes = {
  children: {
    description: 'Content of the coachmark.', // NOTE: Description of this prop.
    control: { type: 'text' }, // NOTE: Control type for this prop.
    table: {
      type: {
        summary: 'ReactNode', // NOTE: Explicit type of this prop.
      },
    },
  },
};

export { coachmarkArgTypes };

export default {
  ...commonStyles,
  ...coachmarkArgTypes,
};
