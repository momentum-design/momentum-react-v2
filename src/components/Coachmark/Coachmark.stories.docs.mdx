The `<Coachmark />` component.

This component should be used with experience-driven flows, such as the "First Time Experience"
[FTE] for a feature. The visibility of each `<Coachmark />` must be controlled by the `isVisible`
prop, and can be dismissed by changing this prop to `false` or if the user clicks the `dismiss`
button. The `dismiss` button triggers the `onDismiss()` event handler.
