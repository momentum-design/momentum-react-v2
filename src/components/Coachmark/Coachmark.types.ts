import type { ReactElement, ReactNode } from 'react';

import { ButtonSimpleProps } from '../ButtonSimple';
import { PopoverProps } from '../Popover';
import { InferredIconName } from '../Icon/Icon.types';

export interface CoachmarkWithoutHeaderProps {
  /**
   * Actions associated with this Coachmark.
   */
  actions?: ReactElement<ButtonSimpleProps> | Array<ReactElement<ButtonSimpleProps>>;

  /**
   * Content to display on this Coachmark.
   */
  children: ReactNode;

  /**
   * Event handler to handle dismissing the Coachmark.
   */
  onDismiss?: () => void;
}

export interface CoachmarkWithHeaderProps extends CoachmarkWithoutHeaderProps {
  /**
   * Icon to display to the left of the title.
   */
  icon?: InferredIconName;

  /**
   * Image associated with this Coachmark.
   */
  image?: ReactElement<HTMLImageElement>;

  /**
   * Title of this Coachmark.
   */
  title?: string;
}

export interface Props extends PopoverProps, CoachmarkWithHeaderProps {
  /**
   * Whether or not this Coachmark is to be visible.
   */
  isVisible?: boolean;
}
