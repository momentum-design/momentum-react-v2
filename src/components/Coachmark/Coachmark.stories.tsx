import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';
import { Story } from '@storybook/react';

import Coachmark, { CoachmarkProps } from './';
import argTypes from './Coachmark.stories.args';
import Documentation from './Coachmark.stories.docs.mdx';
import ButtonPill from '../ButtonPill';
import React from 'react';
import Text from '../Text';

export default {
  title: 'Momentum UI/Coachmark',
  component: Coachmark,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

const Example: Story<CoachmarkProps> = (args: CoachmarkProps) => (
  <div style={{ minHeight: '300px', display: 'inline-block' }}>
    <ButtonPill onPress={() => document.getElementById(args.id)?.showPopover()}>open</ButtonPill>
    <ButtonPill onPress={() => document.getElementById(args.id)?.hidePopover()}>close</ButtonPill>
    <ButtonPill id={args.triggerID}>Trigger (dummy)</ButtonPill>
    <Coachmark {...args} />
  </div>
);

Example.argTypes = { ...argTypes };
Example.args = {
  id: 'coachmark',
  triggerID: 'trigger',
  'close-button-aria-label': 'Close',
  children: (
    <div>
      <Text tagName="p" type="header-primary">
        Coachmark title
      </Text>
      <Text tagName="p" type="body-secondary">
        Some content in here, blah blah.
      </Text>
    </div>
  ),
};

export { Example };
