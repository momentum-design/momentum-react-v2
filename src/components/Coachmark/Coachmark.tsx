import React, { FC } from 'react';

import { Props } from './Coachmark.types';
import { Coachmark as MdcCoachmark } from '@momentum-design/components/dist/react';

/**
 * The Coachmark component.
 */
const Coachmark: FC<Props> = (props: Props) => {
  return <MdcCoachmark {...props} />;
};

export default Coachmark;
