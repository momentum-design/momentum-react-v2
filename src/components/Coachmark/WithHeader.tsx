import React, { FC } from 'react';

import ButtonControl, { BUTTON_CONTROL_CONSTANTS } from '../ButtonControl';
import ButtonGroup from '../ButtonGroup';
import Icon, { ICON_CONSTANTS } from '../Icon';
import Text, { TEXT_CONSTANTS } from '../Text';

import { STYLE } from './Coachmark.constants';
import { CoachmarkWithHeaderProps } from './Coachmark.types';
import './Coachmark.style.scss';

/**
 * The Coachmark content component when displayed with a header.
 */
const CoachmarkWithHeader: FC<CoachmarkWithHeaderProps> = (props: CoachmarkWithHeaderProps) => {
  const { actions, children, icon, image, onDismiss, title } = props;

  return (
    <div className={STYLE.container} data-header="true">
      <div className={STYLE.header}>
        {icon && (
          <Icon data-testid={icon} name={icon} scale={18} weight={ICON_CONSTANTS.WEIGHTS.BOLD} />
        )}
        <Text className={STYLE.title} type={TEXT_CONSTANTS.TYPES.HEADER_PRIMARY} tagName="h3">
          {title}
        </Text>
        <ButtonGroup round>
          <ButtonControl
            aria-label="dismiss"
            control={BUTTON_CONTROL_CONSTANTS.CONTROLS.CLOSE}
            onPress={onDismiss}
          />
        </ButtonGroup>
      </div>
      {image && <div className={STYLE.image}>{image}</div>}
      {children && <div className={STYLE.content}>{children}</div>}
      {actions && (
        <ButtonGroup round spaced>
          {actions}
        </ButtonGroup>
      )}
    </div>
  );
};

export default CoachmarkWithHeader;
