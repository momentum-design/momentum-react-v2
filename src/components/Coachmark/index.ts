import { default as Coachmark } from './Coachmark';
import { Props } from './Coachmark.types';

export type CoachmarkProps = Props;

export default Coachmark;
