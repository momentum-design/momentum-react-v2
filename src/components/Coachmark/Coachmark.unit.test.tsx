import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import ButtonPill from '../ButtonPill';

import Coachmark from './';

jest.mock('uuid', () => {
  return {
    v4: () => '1',
  };
});

describe('<Coachmark />', () => {
  const iconName = 'placeholder';
  const children = 'Children';

  describe('snapshot', () => {
    it('should match snapshot without header', async () => {
      expect.assertions(1);

      const { container } = render(
        <Coachmark
          actions={[
            <ButtonPill key={0}>Button A</ButtonPill>,
            <ButtonPill key={1}>Button B</ButtonPill>,
          ]}
          isVisible
          triggerComponent={<div />}
        >
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with header', async () => {
      expect.assertions(1);

      const { container } = render(
        <Coachmark
          actions={[
            <ButtonPill key={0}>Button A</ButtonPill>,
            <ButtonPill key={1}>Button B</ButtonPill>,
          ]}
          icon={iconName}
          image={
            <img
              alt="example"
              src="https://www.firstbenefits.org/wp-content/uploads/2017/10/placeholder-1024x1024.png"
            />
          }
          isVisible
          title="Title"
          triggerComponent={<div />}
        >
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should allow actions', async () => {
      expect.assertions(2);
      const user = userEvent.setup();

      const onPressMock = jest.fn();

      render(
        <Coachmark
          actions={[
            <ButtonPill key={0} onPress={onPressMock}>
              Button A
            </ButtonPill>,
          ]}
          isVisible
          triggerComponent={<div />}
        >
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      const button = screen.getByRole('button', { name: 'Button A' });

      expect(button).toBeVisible();

      await user.click(button);

      expect(onPressMock).toHaveBeenCalledTimes(1);
    });

    it('should allow icon', async () => {
      expect.assertions(1);

      render(
        <Coachmark isVisible icon={iconName} triggerComponent={<div />}>
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      const icon = screen.getByTestId(iconName);

      expect(icon).toBeVisible();
    });

    it('should allow image', async () => {
      expect.assertions(2);

      const alt = 'example';
      const src =
        'https://www.firstbenefits.org/wp-content/uploads/2017/10/placeholder-1024x1024.png';

      render(
        <Coachmark isVisible image={<img alt={alt} src={src} />} triggerComponent={<div />}>
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      const image = screen.getByAltText<HTMLImageElement>(alt);

      expect(image).toBeVisible();
      expect(image.src).toBe(src);
    });

    it('should allow dismiss via keypress', async () => {
      expect.assertions(3);
      const user = userEvent.setup();

      const mockDismiss = jest.fn();

      render(
        <Coachmark isVisible onDismiss={mockDismiss} triggerComponent={<div />}>
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      const dismissButton = screen.getByRole('button', { name: 'dismiss' });

      expect(dismissButton).toBeVisible();

      await user.tab();
      await user.keyboard('{enter}');

      await waitFor(() => {
        expect(screen.queryByText(children)).not.toBeInTheDocument();
      });

      expect(mockDismiss).toHaveBeenCalledTimes(1);
    });

    it('should allow dismiss via click', async () => {
      expect.assertions(3);
      const user = userEvent.setup();

      const mockDismiss = jest.fn();

      render(
        <Coachmark isVisible onDismiss={mockDismiss} triggerComponent={<div />}>
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      const dismissButton = screen.getByRole('button', { name: 'dismiss' });

      expect(dismissButton).toBeVisible();

      await user.click(dismissButton);

      await waitFor(() => {
        expect(screen.queryByText(children)).not.toBeInTheDocument();
      });

      expect(mockDismiss).toHaveBeenCalledTimes(1);
    });

    it('should respond to isVisible prop', async () => {
      expect.assertions(3);

      const { rerender } = render(
        <Coachmark isVisible triggerComponent={<div />}>
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      expect(screen.getByText(children)).toBeVisible();

      rerender(<Coachmark triggerComponent={<div />}>{children}</Coachmark>);

      await waitFor(() => {
        expect(screen.queryByText(children)).not.toBeInTheDocument();
      });

      rerender(
        <Coachmark isVisible triggerComponent={<div />}>
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      expect(screen.getByText(children)).toBeVisible();
    });

    it('should allow title', async () => {
      expect.assertions(1);

      const title = 'Title';

      render(
        <Coachmark isVisible title={title} triggerComponent={<div />}>
          {children}
        </Coachmark>
      );

      await screen.findByText(children);

      expect(screen.getByText(title)).toBeVisible();
    });
  });
});
