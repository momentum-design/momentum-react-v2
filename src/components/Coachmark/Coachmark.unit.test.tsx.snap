// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Coachmark /> snapshot should match snapshot with header 1`] = `
<div>
  <div
    aria-expanded="true"
  />
  <div
    data-tippy-root=""
    id="tippy-2"
    style="z-index: 9999; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="md-coachmark-wrapper md-modal-container-wrapper"
      data-arrow-orientation="vertical"
      data-color="primary"
      data-elevation="3"
      data-padded="true"
      data-placement="top"
      data-round="50"
    >
      <div
        class="md-coachmark-container"
        data-header="true"
      >
        <div
          class="md-coachmark-header"
        >
          <div
            class="md-icon-wrapper md-icon-auto-scales md-icon-scales md-icon-no-shrink"
          >
            <svg
              class=""
              data-autoscale="false"
              data-scale="18"
              data-test="placeholder"
              data-testid="placeholder"
              fill="currentColor"
              height="100%"
              viewBox="0, 0, 32, 32"
              width="100%"
            />
          </div>
          <h3
            class="md-text-wrapper md-coachmark-title"
            data-type="header-primary"
          >
            Title
          </h3>
          <div
            class="md-button-group-wrapper"
            data-compressed="false"
            data-round="true"
            data-separator="false"
            data-spaced="false"
          >
            <button
              aria-label="dismiss"
              class="md-button-control-wrapper md-button-simple-wrapper"
              type="button"
            >
              <div
                class="md-icon-wrapper md-icon-auto-scales md-icon-scales"
              >
                <svg
                  class=""
                  data-autoscale="true"
                  data-scale="false"
                  data-test="cancel"
                  fill="currentColor"
                  height="100%"
                  viewBox="0, 0, 32, 32"
                  width="100%"
                />
              </div>
            </button>
          </div>
        </div>
        <div
          class="md-coachmark-image"
        >
          <img
            alt="example"
            src="https://www.firstbenefits.org/wp-content/uploads/2017/10/placeholder-1024x1024.png"
          />
        </div>
        <div
          class="md-coachmark-content"
        >
          Children
        </div>
        <div
          class="md-button-group-wrapper"
          data-compressed="false"
          data-round="true"
          data-separator="false"
          data-spaced="true"
        >
          <button
            class="md-button-pill-wrapper md-button-simple-wrapper"
            data-color="primary"
            data-disabled="false"
            data-ghost="false"
            data-grown="false"
            data-outline="false"
            data-shallow-disabled="false"
            data-size="40"
            data-solid="false"
            type="button"
          >
            <span>
              Button A
            </span>
          </button>
          <button
            class="md-button-pill-wrapper md-button-simple-wrapper"
            data-color="primary"
            data-disabled="false"
            data-ghost="false"
            data-grown="false"
            data-outline="false"
            data-shallow-disabled="false"
            data-size="40"
            data-solid="false"
            type="button"
          >
            <span>
              Button B
            </span>
          </button>
        </div>
      </div>
      <div
        class="md-modal-container-arrow-wrapper"
        data-popper-arrow="true"
        id="arrow1"
        style="position: absolute; left: 0px; transform: translate(5px, 0px);"
      >
        <svg
          class="md-modal-arrow-svg"
          data-placement="top"
          height="12"
          viewBox="0 0 24 12"
          width="24"
          xmlns="http://www.w3.org/svg/2000"
        >
          <defs>
            <clippath
              id="modal-arrow-cut-stroke-top"
            >
              <path
                d="m 24 0 l -10 10 q -2 2 -4 0 l -10 -10"
              />
            </clippath>
          </defs>
          <path
            clip-path="url(#modal-arrow-cut-stroke-top)"
            d="m 24 0 l -10 10 q -2 2 -4 0 l -10 -10"
            data-color="primary"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    class="md-popover-backdrop"
  />
</div>
`;

exports[`<Coachmark /> snapshot should match snapshot without header 1`] = `
<div>
  <div
    aria-expanded="true"
  />
  <div
    data-tippy-root=""
    id="tippy-1"
    style="z-index: 9999; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="md-coachmark-wrapper md-modal-container-wrapper"
      data-arrow-orientation="vertical"
      data-color="primary"
      data-elevation="3"
      data-padded="true"
      data-placement="top"
      data-round="50"
    >
      <div
        class="md-coachmark-container"
        data-header="false"
      >
        <div>
          <div
            class="md-coachmark-content"
          >
            Children
          </div>
          <div
            class="md-button-group-wrapper"
            data-compressed="false"
            data-round="true"
            data-separator="false"
            data-spaced="true"
          >
            <button
              class="md-button-pill-wrapper md-button-simple-wrapper"
              data-color="primary"
              data-disabled="false"
              data-ghost="false"
              data-grown="false"
              data-outline="false"
              data-shallow-disabled="false"
              data-size="40"
              data-solid="false"
              type="button"
            >
              <span>
                Button A
              </span>
            </button>
            <button
              class="md-button-pill-wrapper md-button-simple-wrapper"
              data-color="primary"
              data-disabled="false"
              data-ghost="false"
              data-grown="false"
              data-outline="false"
              data-shallow-disabled="false"
              data-size="40"
              data-solid="false"
              type="button"
            >
              <span>
                Button B
              </span>
            </button>
          </div>
        </div>
        <div
          class="md-button-group-wrapper md-coachmark-controls"
          data-compressed="false"
          data-round="true"
          data-separator="false"
          data-spaced="false"
        >
          <button
            aria-label="dismiss"
            class="md-button-control-wrapper md-button-simple-wrapper"
            type="button"
          >
            <div
              class="md-icon-wrapper md-icon-auto-scales md-icon-scales"
            >
              <svg
                class=""
                data-autoscale="true"
                data-scale="false"
                data-test="cancel"
                fill="currentColor"
                height="100%"
                viewBox="0, 0, 32, 32"
                width="100%"
              />
            </div>
          </button>
        </div>
      </div>
      <div
        class="md-modal-container-arrow-wrapper"
        data-popper-arrow="true"
        id="arrow1"
        style="position: absolute; left: 0px; transform: translate(5px, 0px);"
      >
        <svg
          class="md-modal-arrow-svg"
          data-placement="top"
          height="12"
          viewBox="0 0 24 12"
          width="24"
          xmlns="http://www.w3.org/svg/2000"
        >
          <defs>
            <clippath
              id="modal-arrow-cut-stroke-top"
            >
              <path
                d="m 24 0 l -10 10 q -2 2 -4 0 l -10 -10"
              />
            </clippath>
          </defs>
          <path
            clip-path="url(#modal-arrow-cut-stroke-top)"
            d="m 24 0 l -10 10 q -2 2 -4 0 l -10 -10"
            data-color="primary"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    class="md-popover-backdrop"
  />
</div>
`;
